File: auto_permissions/providers/gemini.py
"""Google Gemini REST API connector with OpenAPI schema and HTTP 429 cooldown detection."""
import json
import tempfile
import urllib.request
import urllib.error
from pathlib import Path
from typing import Any, Dict, Optional

from .base import BaseProvider, parse_json_safely, resolve_api_key, is_in_ttl_cooldown, write_ttl_cooldown

class GeminiProvider(BaseProvider):
    """Connects to Google Gemini Flash Lite via generative language REST API."""
    def __init__(
        self,
        api_key: Optional[str] = None,
        model: str = "gemini-flash-lite-latest",
        temperature: float = 0.0,
        timeout: float = 4.5,
    ):
        self.model = model
        self.temperature = temperature
        self.timeout = timeout
        self.api_key = api_key or resolve_api_key("GEMINI_API_KEY", "gemini_api_key")

    @staticmethod
    def _get_cooldown_file() -> Path:
        return Path(tempfile.gettempdir()) / "auto_permissions_gemini_cooldown.json"

    def is_in_cooldown(self) -> bool:
        return is_in_ttl_cooldown(self._get_cooldown_file(), key="cooldown_until")

    def _set_cooldown(self, seconds: int = 60, reason: str = "HTTP 429 Rate Limit") -> None:
        write_ttl_cooldown(self._get_cooldown_file(), seconds, key="cooldown_until", extra={"reason": reason})

    def evaluate(self, system_prompt: str, prompt: str) -> Optional[Dict[str, Any]]:
        if not self.api_key or self.is_in_cooldown():
            return None

        url = f"https://generativelanguage.googleapis.com/v1beta/models/{self.model}:generateContent?key={self.api_key}"
        payload = {
            "contents": [
                {"role": "user", "parts": [{"text": f"{system_prompt}\n\n{prompt}"}]}
            ],
            "generationConfig": {
                "responseMimeType": "application/json",
                "responseSchema": {
                    "type": "OBJECT",
                    "properties": {
                        "decision": {
                            "type": "STRING",
                            "enum": ["allow", "deny", "ask", "force_ask"]
                        },
                        "reason": {"type": "STRING"}
                    },
                    "required": ["decision", "reason"]
                },
                "temperature": self.temperature,
            }
        }
        req = urllib.request.Request(
            url,
            data=json.dumps(payload).encode("utf-8"),
            headers={"Content-Type": "application/json"}
        )
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                data = json.loads(resp.read().decode("utf-8"))
                candidates = data.get("candidates", [])
                if candidates:
                    parts = candidates[0].get("content", {}).get("parts", [])
                    if parts:
                        text = parts[0].get("text", "")
                        return parse_json_safely(text)
        except urllib.error.HTTPError as e:
            if e.code == 429:
                err_body = ""
                try:
                    err_body = e.read().decode("utf-8", errors="replace")
                except Exception:
                    pass
                is_daily = "daily" in err_body.lower() or "limit: 500" in err_body.lower()
                cooldown_sec = 86400 if is_daily else 60
                self._set_cooldown(seconds=cooldown_sec, reason=f"Gemini 429 ({'Daily limit' if is_daily else 'RPM limit'})")
            return None
        except Exception:
            return None
        return None

File: internal/providers/gemini.go
package providers

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// GeminiProvider connects to Google Gemini Generative Language REST API.
type GeminiProvider struct {
	APIKey      string
	Model       string
	Temperature float64
	Timeout     time.Duration
	MaxTokens   int
	client      *http.Client
}

func NewGeminiProvider(apiKey, model string, temp float64, timeout time.Duration, maxTokens int) *GeminiProvider {
	if apiKey == "" {
		apiKey = ResolveAPIKey("GEMINI_API_KEY", "", "")
	}
	if model == "" || model == "auto" || model == "default" {
		model = "gemini-flash-lite-latest"
	}
	if timeout <= 0 {
		timeout = 4500 * time.Millisecond
	}
	if maxTokens <= 0 {
		maxTokens = 512
	}
	return &GeminiProvider{
		APIKey:      apiKey,
		Model:       model,
		Temperature: temp,
		Timeout:     timeout,
		MaxTokens:   maxTokens,
		client:      &http.Client{Timeout: timeout},
	}
}

func (p *GeminiProvider) GetEndpoint() string {
	return "https://generativelanguage.googleapis.com"
}

func (p *GeminiProvider) cooldownFile() string {
	return filepath.Join(os.TempDir(), "auto_permissions_gemini_cooldown.json")
}

func (p *GeminiProvider) IsInCooldown() bool {
	return IsInTTLCooldown(p.cooldownFile(), "cooldown_until")
}

func (p *GeminiProvider) setCooldown(seconds float64, reason string) {
	WriteTTLCooldown(p.cooldownFile(), seconds, "cooldown_until", map[string]interface{}{"reason": reason})
}

func (p *GeminiProvider) Evaluate(systemPrompt, prompt string) (map[string]interface{}, string, error) {
	if p.APIKey == "" || p.IsInCooldown() {
		return nil, "CLOUD", fmt.Errorf("gemini unavailable: missing api key or in cooldown")
	}

	url := fmt.Sprintf("https://generativelanguage.googleapis.com/v1beta/models/%s:generateContent?key=%s", p.Model, p.APIKey)
	payload := map[string]interface{}{
		"contents": []map[string]interface{}{
			{
				"role": "user",
				"parts": []map[string]string{
					{"text": fmt.Sprintf("%s\n\n%s", systemPrompt, prompt)},
				},
			},
		},
		"generationConfig": map[string]interface{}{
			"responseMimeType": "application/json",
			"responseSchema": map[string]interface{}{
				"type": "OBJECT",
				"properties": map[string]interface{}{
					"decision": map[string]interface{}{
						"type": "STRING",
						"enum": []string{"allow", "deny", "ask", "force_ask"},
					},
					"reason": map[string]string{"type": "STRING"},
					"alternatives": map[string]interface{}{
						"type": "ARRAY",
						"items": map[string]interface{}{
							"type": "OBJECT",
							"properties": map[string]interface{}{
								"label":   map[string]string{"type": "STRING"},
								"command": map[string]string{"type": "STRING"},
							},
							"required": []string{"label"},
						},
					},
				},
				"required": []string{"decision", "reason"},
			},
			"temperature":     p.Temperature,
			"maxOutputTokens": p.MaxTokens,
		},
	}

	data, err := json.Marshal(payload)
	if err != nil {
		return nil, "CLOUD", err
	}

	req, err := http.NewRequest("POST", url, bytes.NewReader(data))
	if err != nil {
		return nil, "CLOUD", err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := p.client.Do(req)
	if err != nil {
		return nil, "CLOUD", err
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusTooManyRequests {
		body, _ := io.ReadAll(resp.Body)
		bodyStr := strings.ToLower(string(body))
		isDaily := strings.Contains(bodyStr, "daily") || strings.Contains(bodyStr, "limit: 500")
		cooldownSec := 60.0
		if isDaily {
			cooldownSec = 86400.0
		}
		p.setCooldown(cooldownSec, "HTTP 429 Rate Limit")
		return nil, "CLOUD", fmt.Errorf("gemini rate limited (429)")
	}

	if resp.StatusCode != http.StatusOK {
		return nil, "CLOUD", fmt.Errorf("gemini returned status %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, "CLOUD", err
	}

	var dataResp struct {
		Candidates []struct {
			Content struct {
				Parts []struct {
					Text string `json:"text"`
				} `json:"parts"`
			} `json:"content"`
		} `json:"candidates"`
	}

	if err := json.Unmarshal(body, &dataResp); err == nil && len(dataResp.Candidates) > 0 {
		parts := dataResp.Candidates[0].Content.Parts
		if len(parts) > 0 {
			parsed := ParseJSONSafely(parts[0].Text)
			if parsed != nil {
				return parsed, "CLOUD", nil
			}
		}
	}

	return nil, "CLOUD", fmt.Errorf("failed to parse gemini response")
}
